src/Range/Pattern.php

```diff
@@ -143,20 +153,94 @@ public function getAddressType()
+            switch ($this->getAddressType()) {
+                case AddressType::T_IPv4:
+                    // Default is public
+                    $this->rangeType = RangeType::T_PUBLIC;
+                    $reservedRanges = IPv4::$reservedRanges;
+                    break;
+                case AddressType::T_IPv6:
+                    if (Subnet::fromString('2002::/16')->contains($this)) {
+                        $this->rangeType = Factory::rangeFromBoundaries($this->fromAddress->toIPv4(), $this->toAddress->toIPv4())->getRangeType();
+
+                        return $this->rangeType;
+                    } else {
+                        // Default is public
+                        $this->rangeType = RangeType::T_RESERVED;
+                        $reservedRanges = IPv6::$reservedRanges;
+                    }
+                    break;
+                default:
+                    return;
+            }
+
+            // Check if range is contained within an RFC subnet
+            foreach ($reservedRanges as $reservedRange) {
+                if (Subnet::fromString($reservedRange['cidr'])->contains($this)) {
+                    $this->rangeType = $reservedRange['type'];
+                    break;
+                }
+            }
+
+            // Check if public/reserved range contains an RFC subnet
+            if ($this->rangeType === RangeType::T_PUBLIC || $this->rangeType === RangeType::T_RESERVED) {
+                foreach ($reservedRanges as $reservedRange) {
+                    if ($this->contains(Subnet::fromString($reservedRange['cidr']))) {
+                        if ($this->rangeType !== $reservedRange['type']) {
+                            // RFC 5735 specifies that 255.255.255.255/32 is excluded from 240.0.0.0/4 and 224.0.0.0/4
+                            if ($this->getAddressType() === AddressType::T_IPv4 &&
+                                ($reservedRange['type'] === RangeType::T_LIMITEDBROADCAST ||
+                                 $reservedRange['type'] === RangeType::T_UNSPECIFIED)) {
+                                continue;
+                            }
+
+                            // :: unspecified exists within ::/8 reserved
+                            // ::1/128 loopback exists within ::/8 reserved
+                            if ($this->getAddressType() === AddressType::T_IPv6 &&
+                                ($reservedRange['type'] === RangeType::T_UNSPECIFIED ||
+                                 $reservedRange['type'] === RangeType::T_LOOPBACK)) {
+                                continue;
+                            }
+                            $this->rangeType = null;
+                        }
+                    }
+                }
+            }
+        }
+
+        return $this->rangeType;
+    }
+
     /**
      * {@inheritdoc}
      *
      * @see RangeInterface::contains()
      */
-    public function contains(AddressInterface $address)
```

Comment:
I'd add a new method to work with ranges, for these reasons:
1. if someone extends this class, we have a backward compatibility break (the method signature changes) - see for example what would happen: https://3v4l.org/Edubs
2. keep things as simple as possible: one method to do one thing

Reply:
I agree, a separate function is cleaner.